File: promocode-issue/src/test/java/com/voshodnerd/promocode/issue/PromoCodeIssueApplicationTests.java
package com.voshodnerd.promocode.issue;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.voshodnerd.promocode.issue.repository.CampaignRepository;
import lombok.SneakyThrows;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.springframework.test.web.servlet.MockMvc;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;


@SpringBootTest
@Testcontainers
@AutoConfigureMockMvc
class PromoCodeIssueApplicationTests {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    CampaignRepository campaignRepository;

    @Container
    public static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>(
            "postgres:15-alpine"
    );

    @DynamicPropertySource
    static void configureProperties(DynamicPropertyRegistry registry) {
        registry.add("spring.datasource.url", postgres::getJdbcUrl);
        registry.add("spring.datasource.username", postgres::getUsername);
        registry.add("spring.datasource.password", postgres::getPassword);
        registry.add("spring.liquibase.enabled", () -> true);
        registry.add("spring.liquibase.change-log", () -> "classpath:liquibase/changelog-master.xml");
        registry.add("spring.liquibase.default-schema", () -> "public");
        registry.add("spring.mandatory-file-encoding", () -> "UTF-8");
        registry.add("spring.http.encoding.charset", () -> "UTF-8");
        registry.add("spring.http.encoding.enabled", () -> true);
    }


    @Test
    void contextLoads() {
    }

    @Test
    @SneakyThrows
    void testPerformRequestCampaignList() {
        var result = mockMvc.perform(get("http://localhost:8080/campaign-list"))
                .andExpect(status().is2xxSuccessful()).andReturn();

        ObjectMapper mapper = new ObjectMapper();
        mapper.registerModule(new JavaTimeModule());
        mapper.setDateFormat(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX"));

        assertEquals(result.getResponse().getContentAsString(),
                mapper.writeValueAsString(campaignRepository.findAll()));
    }

    @Test
    public void testFileUpload() throws Exception {
        // Create a mock CSV file
        String campaignId = "1";
        String csvContent = "D12BPUNM\n" +
                "EF3XVUJK\n" +
                "TTZ6CKQF\n" +
                "UKGMT5S5\n" +
                "GS7IOHGP\n" +
                "D4J7WOKB\n" +
                "1UO1XEPR\n" +
                "RRLNZ02Y\n" +
                "IRM5CN6I\n" +
                "HB2E2WRG\n" +
                "V2SIF4NK\n" +
                "0SP8PQDP\n" +
                "MRIC7QM9\n" +
                "ISCZE8B1\n" +
                "EJDF1M45\n" +
                "J1879225\n" +
                "6CV6VJ8O\n" +
                "ISOAWM7J\n" +
                "TAN7IWJT\n" +
                "G26PNZRQ\n";
        MockMultipartFile file = new MockMultipartFile(
                "filename", // name of the @RequestParam in the controller
                "test.csv", // original file name
                "text/csv", // content type
                csvContent.getBytes() // file content
        );
        // Perform the file upload request
        var result = mockMvc.perform(multipart("/upload")
                        .file(file).param("campaignId", campaignId)) // Add campaignId as a request param)
                .andExpect(status().isOk()).andReturn();
        // Assert the response
        assertEquals(result.getResponse()
                .getContentAsString(StandardCharsets.UTF_8), "Загружено");
    }

}

File: promocode-issue-kotlin/src/test/kotlin/com/voshodnerd/promocode/issue/PromocodeIssueApplicationTests.kt
package com.voshodnerd.promocode.issue;

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.voshodnerd.promocode.issue.repository.CampaignRepository
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.mock.web.MockMultipartFile
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.context.DynamicPropertySource
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers
import org.testcontainers.utility.DockerImageName
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat


@SpringBootTest
@Testcontainers
@AutoConfigureMockMvc
class PromoCodeIssueKotlinApplicationTests {

    @Autowired
    private lateinit var mockMvc: MockMvc

    @Autowired
    private lateinit var campaignRepository: CampaignRepository

    companion object {
        @Container
        private val postgres = PostgreSQLContainer(DockerImageName.parse("postgres:15-alpine"))

        @JvmStatic
        @DynamicPropertySource
        fun configureProperties(registry: DynamicPropertyRegistry) {
            registry.add("spring.datasource.url", postgres::getJdbcUrl)
            registry.add("spring.datasource.username", postgres::getUsername)
            registry.add("spring.datasource.password", postgres::getPassword)
            registry.add("spring.liquibase.enabled") { true }
            registry.add("spring.liquibase.change-log") { "classpath:liquibase/changelog-master.xml" }
            registry.add("spring.liquibase.default-schema") { "public" }
        }
    }

    @Test
    fun contextLoads() {
    }

    @Test
    fun testPerformRequestCampaignList() {
        val result = mockMvc.perform(MockMvcRequestBuilders.get("http://localhost:8080/campaign-list"))
            .andExpect(status().is2xxSuccessful())
            .andReturn()

        val mapper = ObjectMapper()
        mapper.registerModule(JavaTimeModule())
        mapper.setDateFormat(SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX"))

        Assertions.assertEquals(
            result.getResponse().getContentAsString(),
            mapper.writeValueAsString(campaignRepository.findAll())
        )
    }

    @Test
    @Throws(Exception::class)
    fun testFileUpload() {
        // Create a mock CSV file
        val campaignId = "1"
        val csvContent = """
             D12BPUNM
             EF3XVUJK
             TTZ6CKQF
             UKGMT5S5
             GS7IOHGP
             D4J7WOKB
             1UO1XEPR
             RRLNZ02Y
             IRM5CN6I
             HB2E2WRG
             V2SIF4NK
             0SP8PQDP
             MRIC7QM9
             ISCZE8B1
             EJDF1M45
             J1879225
             6CV6VJ8O
             ISOAWM7J
             TAN7IWJT
             G26PNZRQ       
             """.trimIndent()
        val file = MockMultipartFile(
            "filename",  // name of the @RequestParam in the controller
            "test.csv",  // original file name
            "text/csv",  // content type
            csvContent.toByteArray() // file content
        )
        // Perform the file upload request
        val result = mockMvc.perform(
            MockMvcRequestBuilders.multipart("/upload")
                .file(file).param("campaignId", campaignId)
        ) // Add campaignId as a request param)
            .andExpect(status().isOk()).andReturn()
        // Assert the response
        Assertions.assertEquals(
            result.response
                .getContentAsString(StandardCharsets.UTF_8), "Загружено"
        )
    }



}
